### PR TITLE
feat: Add scientific notation and dot shorthand support for real literals

### PR DESCRIPTION
**Pull request by fabiomadge**
- Add scientific notation support (e.g., 1.23e5, 5e-2, 123E+4)
- Add trailing dot shorthand (e.g., 1. for 1.0)
- Add leading dot shorthand (e.g., .5 for 0.5)
- Support combining dot shorthands with scientific notation (.125e3)
- Update grammar to handle all new real literal forms
- Add comprehensive tests for parsing and error cases
- Update documentation with consistent terminology

### CODE REVIEW

**Comment by fabiomadge** [comment]
## 📋 Current Status: Ready for Review

### ✅ **Clean Implementation**
This PR has been cleaned up and now contains **only the scientific notation and trailing dot features**. All maintenance/chore changes have been moved to separate PR #6289.

### 🧹 **Separated Concerns**
- **This PR (#6286)**: Pure feature implementation - scientific notation & trailing dots
- **Chore PR (#6289)**: Grammar maintenance - LL1 resolver cleanup

### 🎯 **Final Feature Set**
After extensive development and testing, this PR delivers:
1. **Scientific notation**: `1.23e5`, `5e-2`, `1E+10`
2. **Trailing dot shorthand**: `1.`, `123.`, `1.e2`
3. **Full compatibility**: No conflicts with tuple access, range slicing, etc.

### ✅ **All CI Passing**
- All integration tests pass
- All unit tests pass  
- No breaking changes
- Comprehensive test coverage

### 📚 **Complete Documentation**
- Grammar documentation updated
- Type system documentation enhanced
- Comprehensive test suite included
- News file added for release notes

This PR is now ready for review and provides significant value for scientific computing applications in Dafny! 🚀

**Comment by RustanLeino** [comment]
These syntactic extensions for `real` will be nice!

About the exponents.
* I suggest _not_ supporting an optional `+` in the exponent. So `1e+5` would have to be written `1e5`. Dafny already does not support a unary plus, so I think it's fine not to support such a `+` here.
* I don't like the case-insensitivity of the `e`. I think we should only support a lower-case `e`. I don't see an advantage with providing more ways to write exponents. Moreover, Dafny only supports a lower-case `x` in hexadecimal literals. (Character literals can start with both `'\u` and `'\U`, but they then continue in different ways.)

File: Source/DafnyCore/Dafny.atg [hunk 1]
```diff
@@ -4898,17 +4902,19 @@ Nat<out BigInteger n>
...
-       } catch (System.FormatException) {
-         SemErr(ErrorId.p_bad_decimal_number_format, t, "incorrectly formatted number");
-         d = BaseTypes.BigDec.ZERO;
-       }
-    .)
+= (. d = BaseTypes.BigDec.ZERO; string S = ""; .)
+  ( realnumber                              (. S = Util.RemoveUnderscores(t.val); .)
+  | digits                                  (. var digitToken = t; .)
+    "."                                     (. S = Util.RemoveUnderscores(digitToken.val) + ".0"; .)
+  | "." ( digits | realnumber )             (. S = "0." + Util.RemoveUnderscores(t.val); .)
```
Comment (RustanLeino) [inline Source/DafnyCore/Dafny.atg]: As written, these two cases would allow whitespace before the trailing dot or after the leading dot. That is undesirable. To fix this, do what is done for `!!` around L3577. Something like this:

```suggestion
  | digits                                  (. var digitToken = t; .)
    "."                                     (. S = Util.RemoveUnderscores(digitToken.val) +
                                                   (t.pos == digitToken.pos + digitToken.val.Length ? "" : " ") +
                                                   ".0"; .)
  | "."                                     (. var dotToken = t; .)
    ( digits | realnumber )                 (. S = "0." +
                                                   (t.pos == dotToken + 1 ? "" : " ") +
                                                   Util.RemoveUnderscores(t.val); .)
```

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/ScientificNotationErrors.dfy [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// RUN: %exits-with 2 %resolve "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// Test error cases for scientific notation, trailing-dot shorthand, and leading-dot shorthand
+
+method MalformedScientificNotation() {
```
Comment (RustanLeino) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/ScientificNotationErrors.dfy:6]: Great to have these tests for malformed literals. Add tests where there is whitespace before a trailing dot or after a leading dot.

File: docs/DafnyRef/GrammarDetails.md [hunk 3]
```diff
@@ -1414,7 +1417,7 @@ LiteralExpression =
 
 Nat = ( digits | hexdigits )
 
-Dec = decimaldigits
+Dec = ( realnumber | digits "." | "." ( digits | realnumber ) )
```
Comment (RustanLeino) [inline docs/DafnyRef/GrammarDetails.md:1420]: This is a good way to write it for the reference manual. That is, there is no need to complicate things here by saying that no whitespace is allowed between the digits and the trailing/leading dot -- a reader of the reference manual wouldn't expect such space to be allowed anyhow.

**Review by RustanLeino — CHANGES_REQUESTED** [review]
This is looking great! Just check for whitespace near leading/trailing dot, please.
